tests/test_from_file.py

```diff
@@ -0,0 +1,128 @@
+        config = parse(
+            self.Config, ["--config-file", str((tmp_path / "config.json")), "--a", "7"], load_from_config=True
+        )
+        assert config.a == 7
+        assert config.b == "something"
+        assert config.c is False
+        assert config.z == "dummy"
+
+    def test_nested_with_file_and_override(self, tmp_path: Path) -> None:
+        (tmp_path / "config.yaml").write_text(dump({"a": 6, "s": {"a": 1, "b": "c", "s": {"q": "z"}}}))
+        config = parse(
+            self.Config, ["--config-file", str((tmp_path / "config.yaml")), "--s-b", "d"], load_from_config=True
+        )
+        assert config.a == 6
+        assert config.b == "something"
+        assert config.c is False
+        assert config.z == "dummy"
+        assert config.s.a == 1
+        assert config.s.b == "d"
+        assert config.s.s.q == "z"
+
+    def test_nested_with_prefixed_keys(self, tmp_path: Path) -> None:
+        (tmp_path / "config.yaml").write_text(dump({"a": 6, "s_a": 1, "s": {"b": "c"}}))
+        config = parse(self.Config, ["--config-file", str((tmp_path / "config.yaml"))], load_from_config=True)
+        assert config.a == 6
+        assert config.s.a == 1
+        assert config.s.b == "c"
+
+    def test_nested_with_duplicate_keys(self, tmp_path: Path) -> None:
+        (tmp_path / "config.yaml").write_text(dump({"a": 6, "s_a": 1, "s": {"a": 2, "b": "c"}}))
+        with raises(KeyError) as e:
+            parse(self.Config, ["--config-file", str((tmp_path / "config.yaml"))], load_from_config=True)
+        assert str(e.value) == "'Collision between keys in config file on key s_a.'"
+
+    def test_parse_with_nonexistent_file(self, tmp_path: Path) -> None:
+        with raises(FileNotFoundError):
+            parse(self.Config, ["--config-file", str((tmp_path / "config.json"))], load_from_config=True)
+
+    def test_parse_with_extra_fields(self, tmp_path: Path) -> None:
+        (tmp_path / "config.json").write_text(dumps({"a": 6, "d": "this_is_extra"}))
+        with warns(UserWarning) as warnings:
+            parse(self.Config, ["--config-file", str((tmp_path / "config.json"))], load_from_config=True)
+        assert (
+            str(warnings[0].message) == "The following keys from the provided configuration file were not consumed: d"
+        )
+
+
+class TestParseFromFileNoDefaults:
+    @dataclass
+    class Config:
+        a: int
+        b: str = field(metadata=dict(positional=True))
+        s: SubConfig = field(default_factory=SubConfig)
+        z: str = "dummy"
+
+    def test_parse_with_file_fields_remain_required(self, tmp_path: Path, capsys) -> None:
+        (tmp_path / "config.json").write_text(dumps({"a": 6, "b": "something"}))
+        with raises(SystemExit):
+            parse(self.Config, ["--config-file", str((tmp_path / "config.json"))], load_from_config=True)
+        captured = capsys.readouterr()
+        assert "the following arguments are required: --a, b" in captured.err
```

Comment:
I think this is ambiguous, as Config has properties a, b and SubConfig also has a, b

I think that since the config json has a, b this must be for Config, and the error message should indicate that these a,b that are missing are for SubConfig

Reply:
The test verifies that fields which have no default, as in `Config`, are required fields even when defaults are provided in the file.

Reply:
So the fields are filled by the config, but because the fields have no default, it still raises an error?

I'm not 100% sure I like that behavior, because when you supply both, i.e. 
`--a=5 "something_else" --config-file=config.json` 
I expect the ones that are passed via the CLI to be leading

Reply:
Unfortunately this behaviour is difficult to change (although not impossible). We create an `ArgumentParser` that handles _all_ the arguments - those for the dataclass and the argument for the config file. The arguments for this parser are set as `required=True` if the dataclass field has no default. All of this is done _before_ the arguments are parser and the config file is read, which means that any fields without a default will always be required, regardless of what is in the config file. 

In any case, any argument passed on the CLI always takes precedent over any other value, be it for a field with or without default. For fields without a default, a value always has to be provided on the CLI which means that a corresponding default in the config file is useless.

A potential solution would be to create two `ArgumentParser` instances, where the first only parses the config file argument, and the second is created _with the information from the config file_. Since both parsers then have to ignore certain arguments, we then have to do bookkeeping to warn for unrecognised arguments. All of this gets rather complicated.

Reply:
> Okay let's leave this for now. I think an alternative solution might involve adding all the values from the file to the list of arguments received from the CLI (and skipping duplicates)

That would also require first parsing the arguments (to find the file) and then changing the arguments. The first time around would trigger issues with required arguments if not provided.